feat: add import taxonomy feature [FC-0036]

rpenido opened the pull request:
## Description
This PR adds a new button in the Taxonomy List to allow users to create new taxonomies by importing a CSV/JSON file.

![import-taxonomy](https://github.com/openedx/frontend-app-course-authoring/assets/849463/e80efbd0-3dc8-4c0f-b49d-a4ae678ee794)

It also adds a new menu item in the Taxonomy Card menu and the Taxonomy Details menu to allow users to import tags to existing Taxonomies by importing a CSV/JSON file. This overwrites the current tags.

![import-tags](https://github.com/openedx/frontend-app-course-authoring/assets/849463/5fcd42af-9293-4ff9-8928-e5cbf861280d)

## Supporting Information
* Closes:
  * https://github.com/openedx/modular-learning/issues/115
  * https://github.com/openedx/modular-learning/issues/140
* Depends on:
  * https://github.com/openedx/edx-platform/pull/33663
  * https://github.com/openedx/edx-platform/pull/33752

## Testing instructions
### Import a new taxonomy
1. Have an import taxonomy template ready to be imported. You could get one from below:
    - [template.csv](https://github.com/openedx/frontend-app-course-authoring/files/13322469/template.csv)
    - [template.json](https://github.com/openedx/frontend-app-course-authoring/files/13322474/template.json)
2. Go to the taxonomy list page (http://localhost:2001/taxonomies)
3. Click on the `+ Import` button and select the desired file
4. You will be prompted to input JSON/CSV file, a Taxonomy Name and Description. The Taxonomy Name is required (not blank).
5. Canceling the prompt will abort the operation
6. If the taxonomy is imported successfully, you will see an alert: Taxonomy imported successfully.
7. Importing an invalid taxonomy file, you will see an alert: Import failed - see details in the browser console. The error info will be logged in the browser console.
### Re-import tags to an existing taxonomy
1. Have an import taxonomy template ready to be imported. You could get one from below:
    - [template.csv](https://github.com/openedx/frontend-app-course-authoring/files/13322469/template.csv)
    - [template.json](https://github.com/openedx/frontend-app-course-authoring/files/13322474/template.json)
2. Go to the taxonomy list page (http://localhost:2001/taxonomies)
3. Click on the `Re-import` menu from an existing taxonomy card
4. You will be prompted to select a file to be imported.
5. Canceling the prompt will abort the operation
6. If the taxonomy is imported successfully, you will see an alert: Taxonomy imported successfully.
7. Importing an invalid taxonomy file, you will see an alert: Import failed - see details in the browser console. The error info will be logged in the browser console.
____
Private-ref:
 - [FAL-3532](https://tasks.opencraft.com/browse/FAL-3532)
 - [FAL-3536](https://tasks.opencraft.com/browse/FAL-3536)

openedx-webhooks commented:
Thanks for the pull request, @rpenido! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

pomegranited commented:
@rpenido I think you're missing some style changes from the original PR?  I'm seeing this when I run this branch:

![Screenshot 2023-11-10 110939](https://github.com/openedx/frontend-app-course-authoring/assets/7556571/7ebc6326-22ad-4b3a-84e8-52e5aabbb307)

rpenido commented:
> @rpenido I think you're missing some style changes from the original PR? I'm seeing this when I run this branch:
> 
> ![Screenshot 2023-11-10 110939](https://user-images.githubusercontent.com/7556571/281919277-7ebc6326-22ad-4b3a-84e8-52e5aabbb307.png)

I changed the Dropdown component to make the tests run and broke the layout.
Fixed here: 4515f9f2f56f8b58a1623b7ab525f4a71d645e08

pomegranited commented on src/taxonomy/import-tags/index.js:
```diff
@@ -0,0 +1,5 @@
+import * as actions from './data/actions';
+
+export {
+  actions, // eslint-disable-line import/prefer-default-export
+};
```
It's better to explicitly list what we're exporting here:
```suggestion
// eslint-disable-line import/prefer-default-export
export { importTaxonomy } from './data/actions';
```

And then use it on [TaxonomyListPage](https://github.com/openedx/frontend-app-course-authoring/pull/675/files#diff-60367d61e5f2750d7f4224e061551f5ecf20b56aaddd1ca23ae2852a17f1e424R17) like this:

```javascript
import { importTaxonomy } from './import-tags';
```

rpenido replied:
I was trying to copy the [ADR](https://github.com/openedx/frontend-template-application/blob/master/docs/decisions/0002-feature-based-application-organization.rst) (the reference came from [here](https://jaysoo.ca/2016/02/28/applying-code-organization-rules-to-concrete-redux-code/#module-index-and-constants)), but we will probably only have a second function here. I will change that.

Also, I'll rename `actions` to `utils` to prevent confusion with reducer actions.

pomegranited commented on src/taxonomy/import-tags/data/api.test.js:
```diff
@@ -0,0 +1,53 @@
+  });
+
+  it('should call import taxonomy', async () => {
+    const tags = {
+      tags: [
+        { id: 'tag_1', value: 'Tag 1' },
+      ],
+    };
+    const str = JSON.stringify(tags);
+    const blob = new Blob([str]);
+    const file = new File([blob], 'taxonomy.json', {
+      type: 'application/JSON',
+    });
+    const formData = new FormData();
+    formData.append('taxonomy_name', 'Taxonomy name');
+    formData.append('taxonomy_description', 'Taxonomy description');
+    formData.append('file', file);
+    formData.asymmetricMatch = () => true;
+    axiosMock.onPost(getTaxonomyImportApiUrl(), formData).reply(201, taxonomyImportMock);
```
You can mock all calls to the import endpoint for this test as shown below; no need to construct all this test data.
```suggestion
    axiosMock.onPost(getTaxonomyImportApiUrl()).reply(201, taxonomyImportMock);
```

pomegranited commented on jest.config.js:
```diff
@@ -14,4 +14,5 @@ module.exports = createConfig('jest', {
   moduleNameMapper: {
     '^lodash-es$': 'lodash',
   },
+  modulePathIgnorePatterns: ["<rootDir>/.*/__mocks__"],
```
Why was this needed? (I'm not very familiar with jest configs)

rpenido replied:
I was creating multiple `index.js` files in the various `__mock__` folders, raising a warning that mocks must have unique names. I think removing these `index.js` files will be the proper solution.

pomegranited reviewed (approved):
This looks good @rpenido ! I've left a couple of minor comments which shouldn't be hard to address.

Could you also please update the PR description and un-draft this to get it ready for upstream/CC review?

:+1: pending those changes ^

- [x] I tested this on my devstack while running the branches from the related PRs
- [x] I read through the code
- [x] I checked for accessibility issues by using my keyboard to navigate.
- [x] Includes documentation
- [x] User-facing strings are extracted for translation

pomegranited commented:
@bradenmacdonald This PR is also ready for upstream review -- who should we ask?

bradenmacdonald commented on src/taxonomy/import-tags/data/utils.js:
```diff
@@ -0,0 +1,82 @@
+    * This function is a temporary "Barebones" implementation of the import
+    * functionality with `prompt` and `alert`. It is intended to be replaced
+    * with a component that shows a `ModalDialog` in the future.
+    * See: https://github.com/openedx/modular-learning/issues/116
```
^ Just flagging this in particular in case any reviewers are concerned about `alert()` and `prompt()` :) . This is just a temporary state during development to help get more PRs out faster and more iteratively, and will be replaced before any users ever see this (other than early beta testers).

bradenmacdonald commented:
@jristau1984 another one - could you please line up a review for this PR too? (Or as mentioned, if you folks don't have much capacity for the near future, let me know and I can find a Core Contributor reviewer from Axim or OpenCraft.)

openedx-webhooks commented:
I've created issue [TNL-11201](https://2u-internal.atlassian.net/browse/TNL-11201) in the private 2U Jira.
<!-- data: eyJqaXJhX2lzc3VlcyI6IFt7Im5pY2siOiAiMnUiLCAia2V5IjogIlROTC0xMTIwMSJ9XX0= -->

bradenmacdonald commented:
@rpenido A couple of your PRs will need to be updated with master, including this one.

pomegranited commented:
@rpenido could you add "[FC-0036]" to the PR title here? Thanks :)

pomegranited commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.jsx:
```diff
@@ -0,0 +1,103 @@
+      return menuItems.filter((item) => !['import'].includes(item));
+    }
+    return menuItems;
+  };
```
Don't we also need to check that the user is allowed to import changes to this taxonomy?

pomegranited replied:
Yep absolutely ok to do this in a follow-up task, I've created https://github.com/openedx/modular-learning/issues/160 for this.

rpenido commented:
Hi, @pomegranited! CC @ChrisChV 
I updated the menu code here: f318dfc9c1c143737d449d35e33807ffa292ba99

I'm happy to hear your opinion on whether we could improve this!

rpenido commented:
> I've created issue [TNL-11201](https://2u-internal.atlassian.net/browse/TNL-11201) in the private 2U Jira.

Hi, @jristau1984! How are you doing? :smiley: 

Do we have an estimated due date for this review?

Thank you!

jristau1984 commented:
Due to heavy focus on internal goal delivery before the end of the year, I cannot promise any reviews before January. Tagging @cablaa77 and @jimjohnson8 in case they are able to prioritize this earlier.

rpenido commented on src/taxonomy/TaxonomyListPage.jsx:
```diff
@@ -1,34 +1,44 @@
+  );
+};
 
-  const onDeleteTaxonomy = React.useCallback((id, name) => {
-    deleteTaxonomy({ pk: id }, {
-      onSuccess: async () => {
-        setToastMessage(intl.formatMessage(messages.taxonomyDeleteToast, { name }));
-      },
-      onError: async () => {
-        // TODO: display the error to the user
-      },
-    });
-  }, [setToastMessage]);
```
This code went to taxonomy-menu feature

xitij2000 commented on src/taxonomy/import-tags/data/utils.test.js:
```diff
@@ -0,0 +1,301 @@
+
+      // Capture the onChange handler from the file input element
+      const onChange = mockAddEventListener.mock.calls[0][1];
+      const mockTarget = {
+        target: {
+          files: [
+            'mockFile',
+          ],
+        },
+      };
+
+      onChange(mockTarget);
+
+      return promise;
+    });
+
+    it('should call the api and return error alert', async () => {
+      jest.spyOn(window, 'confirm').mockReturnValueOnce(true);
+      importTags.mockRejectedValue(new Error('test error'));
+
+      const promise = importTaxonomyTags(1, intl).then(() => {
+        expect(importTags).toHaveBeenCalledWith(1, 'mockFile');
+      });
+
+      // Capture the onChange handler from the file input element
+      const onChange = mockAddEventListener.mock.calls[0][1];
+      const mockTarget = {
+        target: {
+          files: [
+            'mockFile',
+          ],
+        },
+      };
+
+      onChange(mockTarget);
+
+      return promise;
+    });
+  });
+});
```
Additionally, I think some of the code here is quite repetitive. Here are a few suggestions:

- I see repetition of `const onChange ...` this could be moved to a utility function like: `mockOnChange` that cane get take the files array as a parameter. 
- OR. Even better, just mock the `selectFile` function with a known return value. You can test the `selectFile` function separately. This should also clean up and simplify a lot of this code.

rpenido replied:
I expect that this whole file will be replaced. The `importTaxonomyTags` was already replaced in https://github.com/open-craft/frontend-app-course-authoring/pull/19 ([here](https://github.com/open-craft/frontend-app-course-authoring/pull/19/files#r1428382384)). So, I went for the `/* istanbul ignore next */ ` route to save time. Let me know if you are ok with this!

xitij2000 replied:
Sure, that makes sense!

xitij2000 reviewed (commented):
These are mostly nits. It's looking good overall. 
I just need to test it once.

rpenido commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.test.jsx:
```diff
@@ -0,0 +1,251 @@
+      const { getByTestId } = render(<TaxonomyMenuComponent iconMenu={iconMenu} />);
+
+      // Menu closed/doesn't exist yet
+      expect(() => getByTestId('taxonomy-menu')).toThrow();
```
Thank you @xitij2000! That's precisely what we want here.
Fixed: dddd44127d4c18a351778b3b3634ad07b6c13879

rpenido commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.test.jsx:
```diff
@@ -0,0 +1,251 @@
+    jest.clearAllMocks();
+  });
+
+  [true, false].forEach((iconMenu) => {
```
Done: a8b419792ac19d1238819c7c7f57a0fe04eee342

rpenido commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.jsx:
```diff
@@ -0,0 +1,147 @@
+    return menuItems;
+  };
+
+  const menuItems = getTaxonomyMenuItems();
```
There is no need to be a function here. Refactored: a3c40fd36efc41c20a519f9da1a0cb509b1777e1
Also added types and fixed a bug

rpenido commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.jsx:
```diff
@@ -0,0 +1,147 @@
+  IconButton,
+} from '@edx/paragon';
+import { MoreVert } from '@edx/paragon/icons';
+import _ from 'lodash';
```
Done a3c40fd36efc41c20a519f9da1a0cb509b1777e1

rpenido commented on src/taxonomy/TaxonomyListPage.test.jsx:
```diff
@@ -102,20 +105,21 @@ describe('<TaxonomyListPage />', async () => {
-      callbacks.onSuccess();
+    await act(async () => {
+      const { getByTestId } = render(<RootWrapper />);
+      const importButton = getByTestId('taxonomy-import-button');
+      expect(importButton).toBeInTheDocument();
+      importButton.click();
+      expect(importTaxonomy).toHaveBeenCalled();
     });
-    const { getByTestId, getByLabelText } = render(<RootWrapper />);
-    fireEvent.click(getByTestId('test-delete-button'));
-    fireEvent.change(getByLabelText('Type DELETE to confirm'), { target: { value: 'DELETE' } });
-    fireEvent.click(getByTestId('delete-button'));
-
-    expect(mockDeleteTaxonomy).toBeCalledTimes(1);
-    expect(mockSetToastMessage).toBeCalledWith(`"${taxonomies[0].name}" deleted`);
```
This is only for code coverage. Is it ok to skip it?

rpenido commented on src/taxonomy/import-tags/data/api.test.js:
```diff
@@ -0,0 +1,48 @@
+  getTagsImportApiUrl,
+  importNewTaxonomy,
+  importTags,
+} from './api';
+
+let axiosMock;
+
+describe('import taxonomy api calls', () => {
+  beforeEach(() => {
+    initializeMockApp({
+      authenticatedUser: {
+        userId: 3,
+        username: 'abc123',
+        administrator: true,
+        roles: [],
+      },
+    });
+    axiosMock = new MockAdapter(getAuthenticatedHttpClient());
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should call import new taxonomy', async () => {
+    axiosMock.onPost(getTaxonomyImportNewApiUrl()).reply(201, taxonomyImportMock);
+    const result = await importNewTaxonomy('Taxonomy name', 'Taxonomy description');
+
+    expect(axiosMock.history.post[0].url).toEqual(getTaxonomyImportNewApiUrl());
+    expect(result).toEqual(taxonomyImportMock);
+  });
+
+  it('should call import tags', async () => {
+    axiosMock.onPut(getTagsImportApiUrl(1)).reply(200, tagImportMock);
+    const result = await importTags(1);
+
+    expect(axiosMock.history.put[0].url).toEqual(getTagsImportApiUrl(1));
+    expect(result).toEqual(tagImportMock);
+  });
+});
```
I follow this pattern throughout the code. I would like your opinion on my rationale before changing this.
With the [ADR0002](https://github.com/openedx/frontend-template-application/blob/master/docs/decisions/0002-feature-based-application-organization.rst) in mind, I tried to isolate the features in the tests to make it "self-contained".

In this particular case, the UI is in the `taxonomy-menu` feature and the actual import code is in the `taxonomy-import`.
So, I wrote all the tests with these abstractions in mind:
- I test if the UI calls a function (or shows a dialog) that will start a flow from another feature and handle its results (like calling the function to show a toast)
- I test that this function/dialog works and calls the appropriate hooks/apis
- If there is some API hook in the feature, I wrote a unit test to check if this hook makes the correct API function call and side effects (i.e., invalidate other queries)
- And finally, I check if the API function calls make the correct HTTP request with the correct parameters

I saw comments in other PRs wanting to make these tests more "combined" or "user-oriented", using the UI and checking the Axios calls. I understand that, but I found it difficult to implement without having to:
- Break the "feature barrier"
- Duplicating these tests if we reuse some api/hook/feature in more than one place
- Making the tests too complex (the same tests check the UI components, the HTTP calls and responses, react-query invalidations, toast show, alert, etc..)
- Making the tests too coupled and hard to update

The route I took, basically ended with almost all module files with their own test files for their functions, without much integration. This simplifies some code changes on our side:
   - If we change the import flow, we just need to check that the `taxonomy-menu` feature is calling the correct function/component, and we can abstract the actual implementation of it
   - If we change the API, I just need to rewrite the API Tests, not the function/component that prepares the data for it
   - If we change the way we show success/error messages we just need to update the code in the Layout that handles it; not all functions that show success/error messages

We have pros/cons from both approaches, so I will happily follow your lead on the best way to do this for this project: combining the tests, a more unitary approach or a compromise between the two.

xitij2000 replied:
I understand your concerns. I'll highlight my reasons for recommending testing in a more combined way and perhaps @bradenmacdonald can also give his views on the same. 

- Re: breaking the "feature barrier". If you're importing one feature from another, you've already broken it. If you have three components, A, B and C, where B and C are under A, and B and C are importing from each other, then perhaps you can test at the level of A. Or you can look at reorganising so that there is no cross-importing. If both B and C are using the same component then perhaps those components should be moved to a common location from which both can import the component. 
- Duplicating tests: I think this is a good thing. More on this later.
- More complex tests: I don't think the tests will be more complex. I think for the API responses you can define the responses once and then reuse those across multiple tests. 
- Making the tests too coupled: The tests are coupled if your code is coupled, not otherwise. 

I think one big disadvantage of testing thing in isolation is that these are not isolated components. If you are testing 5 different components and they all include some common components, then it might seem like those common components are being tested 5 times, but I think they *should* be tested 5 times!

Here is an example. You create a shared component, let's call it TagList. This is now used by both TagListDropdown and TagListModal. At one point you are working on TagListModal and need to make changes to the core TagList component. You update it and TagListModal and they work great together. The TagList tests pass because they only test TagList. The TagListModal and TagListDropdown tests also pass because they both only test their component and mock the shared TagList component. In actuality if you were to go to a page that uses TagListDropdown, it would break, because the changes to TagList made them incompatible. If you used more integrated tests, that didn't use mocks then the tests would ideally fail in this scenario. 

I think if two "features" of the app are completely isolated then it's definitely OK to have them be isolated in tests. However, the fact that the UI is in taxonomy-menu and the import code is in taxonomy-import is exactly the reason why they should be tested together, because they might both work and pass tests in isolation and be completely broken in practice.

rpenido replied:
Hi @xitij2000! Thank you for taking the time to provide a detailed response!

> If you're importing one feature from another, you've already broken it.
I think it is okay to import from siblings from the ADR. I was concerned about putting the tests of feature A in feature B, but I got the overall idea. Maybe the `taxonomy-menu` here is not the best example (because it's just a menu and unlikely to fail), but your TagList example makes a lot of sense.

I will work to write the test this way for the following features.

Let me know if I should fix it here (and in the other PRs). I'm concerned about the budget, but if you think it is worth it, I'm okay with rewriting the tests to improve them.

Again, thank you for the explanation!

bradenmacdonald replied:
I tend to agree with @xitij2000 's overall philosophy here. I personally tend to write tests that use (coupled) components exactly as they are used in the app, because (A) it can find bugs with the actual integration of the components, like @xitij2000 explained with that example, (B) it's often much easier to do so than to figure out how to use mocks to isolate the components, and (C) I've spent a lot of time over the years hunting down and fixing test failures that were caused by the mocks, not the actual components themselves.

bradenmacdonald commented:
@rpenido  @xitij2000 Are we close to getting this merged? And the other two open PRs? We're quite behind schedule on this PR now.

xitij2000 commented:
> @rpenido @xitij2000 Are we close to getting this merged? And the other two open PRs? We're quite behind schedule on this PR now.

There seems to be one failing test. Once that is fixed this looks good to merge.

xitij2000 commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.test.jsx:
```diff
@@ -0,0 +1,250 @@
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
+import { render, fireEvent, waitFor } from '@testing-library/react';
+import PropTypes from 'prop-types';
+import each from 'jest-each';
```
`jest-each` is included in jest, so you don't need to install the package or import is. Just do `describe.each` instead of `each.describe`

xitij2000 commented on src/taxonomy/taxonomy-menu/TaxonomyMenu.test.jsx:
```diff
@@ -0,0 +1,250 @@
+  allowFreeText: false,
+};
+
+each([true, false]).describe('<TaxonomyMenu iconMenu=%s />', async (iconMenu) => {
```
```suggestion
describe.each([true, false])('<TaxonomyMenu iconMenu=%s />', async (iconMenu) => {
```

xitij2000 commented on package.json:
```diff
@@ -93,6 +93,7 @@
     "glob": "7.2.3",
     "husky": "7.0.4",
     "jest-canvas-mock": "^2.5.2",
+    "jest-each": "^29.7.0",
```
```suggestion
```

xitij2000 reviewed (approved):
One final small change and this is good to merge.

xitij2000 commented:
> One final small change and this is good to merge.

To avoid another review pass I've just applied this change and will merge when tests pass.

openedx-webhooks commented:
@rpenido 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/675&entry.1484655318=https://github.com/rpenido&entry.752974735=2023-11-08+20:45&entry.1917517419=2024-01-08+07:08&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

rpenido commented:
> To avoid another review pass I've just applied this change and will merge when tests pass.

Thank you @xitij2000!
I will update the other PRs with these changes.
